File: clients/dart/chex.dart
// CHEX client — drives the `chex` binary's persistent NDJSON loop.
//
// Dart SDK only (dart:io, dart:convert). Requires the `chex` binary on PATH or
// an explicit path. One long-lived subprocess.
//
//   final c = await CHEX.open();
//   try {
//     final data = await c.validate('./schemas/person.schema.json', {'name': 'Jane Doe', 'age': 30});
//     // name form, resolved against a directory:
//     await c.validate('person', {'name': 'Jane Doe', 'age': 30}, schemaDir: './schemas');
//   } finally {
//     await c.close();
//   }
//
// `validate` returns the validated data and throws CHEXException on a schema
// mismatch. Method names follow Dart's camelCase. `request` is a raw escape
// hatch returning the full response map. Requests are queued: each completes
// with its own response line, in order.

import 'dart:async';
import 'dart:convert';
import 'dart:io';

class CHEXException implements Exception {
  final String message;
  CHEXException(this.message);
  @override
  String toString() => 'CHEXException: $message';
}

class CHEX {
  final Process _proc;
  final _queue = <Completer<Map<String, dynamic>>>[];

  CHEX._(this._proc) {
    _proc.stdout
        .transform(utf8.decoder)
        .transform(const LineSplitter())
        .listen((line) {
      if (line.trim().isEmpty) return;
      if (_queue.isEmpty) return;
      _queue.removeAt(0).complete(jsonDecode(line) as Map<String, dynamic>);
    }, onDone: () {
      final error = CHEXException('chex closed the stream (stderr may have details)');
      while (_queue.isNotEmpty) {
        _queue.removeAt(0).completeError(error);
      }
    });
  }

  /// Start a warm chex process. [binary] defaults to "chex".
  static Future<CHEX> open([String binary = 'chex']) async {
    final proc = await Process.start(binary, ['exec', '--loop']);
    return CHEX._(proc);
  }

  /// Send one raw machine-protocol op; complete with the full response map.
  Future<Map<String, dynamic>> request(Map<String, dynamic> op) {
    final completer = Completer<Map<String, dynamic>>();
    _queue.add(completer);
    _proc.stdin.writeln(jsonEncode(op));
    return completer.future;
  }

  /// Validate [data] against a schema (name or .schema.json path). Returns the validated data.
  Future<dynamic> validate(String schema, Map<String, dynamic> data, {String? schemaDir}) async {
    final op = <String, dynamic>{'op': 'validate', 'schema': schema, 'data': data};
    if (schemaDir != null) op['schemaDir'] = schemaDir;
    final response = await request(op);
    if (response['ok'] != true) {
      final error = response['error'];
      throw CHEXException(error is Map ? (error['message'] ?? 'chex error') : 'chex error');
    }
    return response['result'];
  }

  /// Close stdin so the loop ends, and wait for the process to exit.
  Future<void> close() async {
    await _proc.stdin.close();
    await _proc.exitCode;
  }
}

File: clients/csharp/Chex.cs
// CHEX client — drives the `chex` binary's persistent NDJSON loop.
//
// No NuGet dependencies (System.Text.Json ships with .NET). Requires the `chex`
// binary on PATH or an explicit path. One long-lived subprocess.
//
//   using var c = new Chex();
//   JsonElement data = c.Validate("./schemas/person.schema.json", new { name = "Ada" });
//   JsonElement data2 = c.Validate("person", new { name = "Ada" }, "./schemas");
//
// `Validate` returns the validated data as a JsonElement and throws
// ChexException when it does not match the schema. Method names follow .NET
// PascalCase. Request(json) is the raw escape hatch returning the full response.

using System;
using System.Diagnostics;
using System.Text.Json;

namespace Chex
{
    public sealed class ChexException : Exception
    {
        public ChexException(string message) : base(message) { }
    }

    public sealed class Chex : IDisposable
    {
        private readonly Process _proc;
        private readonly object _lock = new object();

        public Chex(string binary = "chex")
        {
            var psi = new ProcessStartInfo
            {
                FileName = binary,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
            };
            psi.ArgumentList.Add("exec");
            psi.ArgumentList.Add("--loop");
            _proc = Process.Start(psi) ?? throw new InvalidOperationException("failed to start chex");
        }

        /// <summary>Send one raw machine-protocol op (JSON string); returns the full response.</summary>
        public JsonDocument Request(string opJson)
        {
            lock (_lock) // ponytail: one call in flight; drop the lock only if you pipeline
            {
                if (_proc.HasExited) throw new InvalidOperationException("chex process has exited");
                _proc.StandardInput.Write(opJson.TrimEnd());
                _proc.StandardInput.Write('\n');
                _proc.StandardInput.Flush();
                string? line = _proc.StandardOutput.ReadLine();
                if (line == null) throw new InvalidOperationException("chex closed the stream");
                return JsonDocument.Parse(line);
            }
        }

        // Send a fully-formed op JSON and return `result`, throwing on failure.
        private JsonElement Op(string opJson)
        {
            using JsonDocument doc = Request(opJson);
            JsonElement root = doc.RootElement;
            if (!root.GetProperty("ok").GetBoolean())
            {
                string msg = root.TryGetProperty("error", out var e) &&
                             e.TryGetProperty("message", out var m)
                    ? m.GetString() ?? "chex error"
                    : "chex error";
                throw new ChexException(msg);
            }
            return root.TryGetProperty("result", out var r) ? r.Clone() : default;
        }

        /// <summary>Validate data against a schema (name or .schema.json path). Returns the validated data.</summary>
        public JsonElement Validate(string schema, object data, string? schemaDir = null)
        {
            var op = new
            {
                op = "validate",
                schema,
                data,
                schemaDir,
            };
            // Omit null schemaDir so a name resolves the same as on the CLI.
            var opts = new JsonSerializerOptions { DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull };
            return Op(JsonSerializer.Serialize(op, opts));
        }

        public void Dispose()
        {
            if (!_proc.HasExited)
            {
                _proc.StandardInput.Close(); // EOF ends the loop
                _proc.WaitForExit(30_000);
            }
            _proc.Dispose();
        }
    }
}
